File: 백준/Platinum/8986. 전봇대/전봇대.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio                        \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL)
#define ll long long
ll arr[100005];
int n;
ll solve(ll x)
{
    ll ret = 0;
    for (int i = 0; i < n; i++)
    {
        ret += abs(i * x - arr[i]);
    }
    return ret;
}

int main()
{
    fastio;
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    ll l = 1;
    ll r = 1e9;
    while (r - l >= 3)
    {
        ll a = (l * 2 + r) / 3;
        ll b = (l + 2 * r) / 3;
        if (solve(a) > solve(b))
            l = a;
        else
            r = b;
    }
    ll ret = LLONG_MAX;
    for (int i = l; i <= r; i++)
    {
        ret = min(ret, solve(i));
    }
    cout << ret;
    return 0;
}

File: 백준/Platinum/8986. 전봇대/전봇대.py
import math
n = int(input())
arr = [*map(int, input().split())]

def solve(x):    
    ret = 0
    for i in range(n):
        ret += abs(arr[i] - i * x)
    return ret

l ,r = 1, int(1e9)
while r - l >= 3:
    ml = int((l*2+r)/3)
    mr = int((l+2*r)/3)
    if(solve(ml) > solve(mr)):
        l = ml
    else:
        r = mr
ret = float('inf')
for i in range(l, r+1):
    ret = min(ret, solve(i))
print(ret)


        